docs/architecture/adr-014-improve-permissions.md

```diff
@@ -0,0 +1,202 @@
+}
+```
+
+#### `UserPermission`
+
+Instead of storing user permissions as follows: 
+
+```
+UserPermissionPrefix | SubspaceID | User | -> uint32 
+```
+
+We will define a new `UserPermission` object: 
+
+```protobuf
+syntax = "proto3";
+
+// UserPermission contains the details of a single user permissions withing a subspace
+message UserPermission {
+  // Address of the user
+  bytes user = 1;
+  
+  // Permissions set to the user
+  repeated string permissions = 2;
+}
+```
+
+then the permissions of a user will be stored as follows: 
+
+```
+UserPermissionPrefix | SubspaceID | User | -> ProtcolBuffer(UserPermission)
+```
+
+### `Msg` Service
+We will change the `Msg` service of the `x/subspaces` module to properly allow setting permissions: 
+
+```protobuf
+syntax = "proto3";
+
+// MsgSetUserGroupPermissions represents the message used to set the permissions of a user group
+message MsgSetUserGroupPermissions {
+  uint64 subspace_id = 1;
+  uint32 group_id = 2;
+  repeated string permissions = 3;
+  string signer = 4;
+}
+
+// MsgSetUserPermissions represents the message used to set the permissions of a specific user
+message MsgSetUserPermissions {
+  uint64 subspace_id = 1;
+  string user = 2;
+  repeated string permissions = 3;
+  string signer = 4;
+}
+```
+
+#### `Query` Service
+
+We will change the `Query` service of the `x/subspaces` module to properly returns new permission values:
+
+```protobuf
+syntax = "proto3";
+
+// QueryUserPermissionsRequest is the response type for the Query/UserPermissions method
+message QueryUserPermissionsResponse {
+  repeated string permissions = 1;
+  repeated PermissionDetail details = 2;
+}
+```
+
+## Consequences
+
+### Backwards Compatibility
+
+The above detailed solution is **not** backward compatible. For this reason, we will need to write a migration code that reads all the currently set permissions and performs the following operations: 
+- split the permission value into individual permissions;
+- map each individual permission to the new `Permission` type; 
+- replace the current store values with the new `Permission` values. 
+
+This will need to be performed for all user groups as well as all individual user permissions. 
+
+### Positive
+
+- Allows to define permission in all modules
+- No need to sanitize the values received from clients
+- Easier permission setting from clients (values are now human-readable)
+
+### Negative
+
+- Storing `string` instead of `uint` takes up more storage space  
+
+### Neutral
+
+
+## Further Discussions
+
+In the future, we might even allow developers to register custom permissions within their own subspaces, and then request those permissions during the execution of different messages. This could be done by allowing them to specify a `MsgTypeURL -> []Permission` entry for each kind of message type, so that when a message with `MsgTypeURL` gets executed, the user needs to have the specified permission in order to successfully perform the request.
```

Comment:
Does it assume the subspace is controlled by contract, and the feature is about registering the permission for contract?

Reply:
No, the idea is to allow any subspace owner to set a subspace-specific set of permissions to be checked. This might be useful if they want some kind of custom permission to be set on a user before they can do something. An example might be allowing only specific users to interact with smart contracts inside their subspace (e.g. paying users) or anything else.

Anyway, this is there only as a reminder for future iterations that we might want to implement.